torch/_dynamo/variables/higher_order_ops.py

```diff
@@ -889,6 +891,106 @@ def _check_supported_callable(fn_var):
         )
 
 
+class AssociativeScanHigherOrderVariable(TorchHigherOrderOperatorVariable):
+    @raise_hard_error_if_graph_break(
+        reason="associative_scan must be captured completely with torch.compile."
+    )
+    def call_function(
+        self, tx, args: List[VariableTracker], kwargs: Dict[str, VariableTracker]
+    ) -> VariableTracker:
+        from . import NestedUserFunctionVariable, UserFunctionVariable
+        from .builder import SourcelessBuilder, wrap_fx_proxy
+
+        args, kwargs = LazyVariableTracker.realize_all((args, kwargs))
+
+        def arg_extractor(combine_fn, input, dim):
+            return combine_fn, input, dim
+
+        combine_fn, input, dim = arg_extractor(*args, **kwargs)
+
+        assert isinstance(
+            combine_fn,
+            (
+                UserFunctionVariable,
+                NestedUserFunctionVariable,
+            ),
+        )
+
+        if input.python_type() != torch.Tensor:
+            unimplemented(
+                f"Expected input to be a tensor but got {input.python_type()}",
+            )
+
+        # Trace the subgraph
+        # TODO: Fix these pointless new_empty calls appearing in the dynamo output graph.
+        null_shape = SourcelessBuilder.create(tx, ())
+        sub_args = [
+            input.call_method(tx, "new_empty", args=(null_shape,), kwargs={})
+            for _ in range(2)
+        ]
+        (
+            (combine_result, combine_treespec),
+            combine_graph,
+            combine_lifted_freevars,
+        ) = speculate_subgraph(
+            tx,
+            combine_fn,
+            sub_args,
+            sub_kwargs={},
+            description="scan_combine",
+            source_target=self.value,
+            set_subgraph_inputs="flatten_manual",
+        )
+
+        if combine_lifted_freevars:
+            unimplemented(
+                f"Combine fn had unexpected freevars: {combine_lifted_freevars}"
+            )
+
+        if combine_result.python_type() != torch.Tensor:
+            unimplemented(
+                f"Expected combine_fn to return a tensor but got {combine_result.python_type()}",
+            )
+
+        input_meta = input.as_proxy().node.meta["example_value"]
+        combine_result_meta = combine_result.as_proxy().node.meta["example_value"]
+
+        if combine_result_meta.device != input_meta.device:
+            unimplemented(
+                f"Expected combine_fn to return a tensor on device {input_meta.device} but "
+                + f"got {combine_result_meta.device}"
+            )
+        if combine_result_meta.dtype != input_meta.dtype:
+            unimplemented(
+                f"Expected combine_fn to return a tensor of {input_meta.dtype} but "
+                + f"got {combine_result_meta.dtype}"
+            )
+
+        if combine_result_meta.shape != ():
```

Comment:
It'd be great to have a very weak check with `atol=0.1` or so, where we check that the function is associative using the first 3 elements of the tensor (if it has at least 3 in `dim`). This would be *super* useful to catch bugs when implementing non-trivial scans.

Reply:
I was thinking where we would put this and there are a few possibilities:
1. Do it on a best effort basis: Create a 0-D tensor, potentially on CPU, and to assert_close. If there is any error when executing the ops, just skip the test.
2. Hide it behind a debug flag and advertise the flag in the docs. Then when this flag is on, we would realize the inputs to the scan and add an assert in the wrapper code. Otherwise, we could write a device_assert inside the code (provided slices already work in triton).

Probably option 1 (which we could also hide behind a flag, if you think that we may hit false positives?) may be already fine and would catch enough errors early on, being able to throw a nice error.